Mage/src/main/java/mage/abilities/effects/common/GainsChoiceOfAbilitiesEffect.java

```diff
@@ -0,0 +1,120 @@
+
+package mage.abilities.effects.common;
+
+import mage.abilities.Ability;
+import mage.abilities.Mode;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.common.continuous.GainAbilityTargetEffect;
+import mage.choices.Choice;
+import mage.choices.ChoiceImpl;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.targetpointer.FixedTargets;
+
+import java.util.*;
+
+/**
+ * Something gains one ability of the player's choice from a list of abilities until end of turn.
+ * Affects any targets set and optionally the source as well. Have no targets to affect only the source.
+ *
+ * @author notgreat, TheElk801
+ */
+public class GainsChoiceOfAbilitiesEffect extends OneShotEffect {
+
+    private final Map<String, Ability> abilityMap = new LinkedHashMap<>();
+    private final boolean affectSource, includeEnd;
```

Comment:
named like additionallyAffectSource or add comments about it (that it gives effects to target permanents and to a source).

includeEnd - includeEndText

Reply:
The majority of uses of that will affect only the source, and no target will be set. See [[Golem Artisan]] and the like.

Reply:
Sorry, that's a bad example, that one targets. Lunar Avenger and others in https://scryfall.com/search?q=o%3A%7E+o%3A%22gains+your+choice%22+f%3Av

Reply:
BTW if it works for source only then better named as xxxSourceEffect.